components/ws-manager/pkg/manager/manager.go

```diff
@@ -346,6 +361,117 @@ func (m *Manager) StartWorkspace(ctx context.Context, req *api.StartWorkspaceReq
+	// restore is a two step process
+	// 1. create volume snapshot referencing not yet created volume snapshot content
+	// 2. create volume snapshot content referencing volume snapshot
+	// this creates correct bidirectional binding between those two
+
+	// todo(pavel): figure out if there is a way to find out which snapshot class we need to use here. For now use default class info.
+	var volumeSnapshotClassName string
+	if m.Config.WorkspaceClasses[config.DefaultWorkspaceClass].PVC.SnapshotClass != "" {
+		volumeSnapshotClassName = m.Config.WorkspaceClasses[config.DefaultWorkspaceClass].PVC.SnapshotClass
+	}
+
+	var volumeSnapshotClass volumesnapshotv1.VolumeSnapshotClass
+	err = m.Clientset.Get(ctx, types.NamespacedName{Namespace: "", Name: volumeSnapshotClassName}, &volumeSnapshotClass)
+	if err != nil {
+		return fmt.Errorf("was unable to get volume snapshot class: %v", err)
+	}
+
+	snapshotContentName := "restored-" + id
+	volumeSnapshot := &volumesnapshotv1.VolumeSnapshot{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      id,
+			Namespace: m.Config.Namespace,
+		},
+		Spec: volumesnapshotv1.VolumeSnapshotSpec{
+			Source: volumesnapshotv1.VolumeSnapshotSource{
+				VolumeSnapshotContentName: &snapshotContentName,
+			},
+		},
+	}
+	if volumeSnapshotClassName != "" {
+		volumeSnapshot.Spec.VolumeSnapshotClassName = &volumeSnapshotClassName
+	}
+
+	err = m.Clientset.Create(ctx, volumeSnapshot)
+	if err != nil && !k8serr.IsAlreadyExists(err) {
+		return fmt.Errorf("cannot create volumesnapshot: %v", err)
+	}
+
+	volumeSnapshotContent := &volumesnapshotv1.VolumeSnapshotContent{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      snapshotContentName,
+			Namespace: "", // content is not namespaced
+		},
+		Spec: volumesnapshotv1.VolumeSnapshotContentSpec{
+			VolumeSnapshotRef: corev1.ObjectReference{
+				Kind:      "VolumeSnapshot",
+				Name:      id,
+				Namespace: m.Config.Namespace,
+			},
+			DeletionPolicy: "Delete",
+			Source: volumesnapshotv1.VolumeSnapshotContentSource{
+				SnapshotHandle: &handle,
+			},
+			Driver: volumeSnapshotClass.Driver,
+		},
+	}
+
+	err = m.Clientset.Create(ctx, volumeSnapshotContent)
+	if err != nil && !k8serr.IsAlreadyExists(err) {
+		return fmt.Errorf("cannot create volumesnapshotcontent: %v", err)
+	}
+
+	return nil
+}
+
+func (m *Manager) DeleteVolumeSnapshot(ctx context.Context, req *api.DeleteVolumeSnapshotRequest) (res *api.DeleteVolumeSnapshotResponse, err error) {
+	span, ctx := tracing.FromContext(ctx, "DeleteVolumeSnapshot")
+	tracing.LogRequestSafe(span, req)
+	defer tracing.FinishSpan(span, &err)
+	log := log.WithField("func", "DeleteVolumeSnapshot")
+
+	okResponse := &api.DeleteVolumeSnapshotResponse{}
+
+	var volumeSnapshot volumesnapshotv1.VolumeSnapshot
+	err = m.Clientset.Get(ctx, types.NamespacedName{Namespace: m.Config.Namespace, Name: req.Id}, &volumeSnapshot)
+	if k8serr.IsNotFound(err) {
+		if !req.SoftDelete {
+			err = m.restoreVolumeSnapshotFromHandle(ctx, req.Id, req.VolumeHandle)
```

Comment:
Not quite understand that _why_ do we need to support soft delete?
The purpose seems like we will restore the VolumeSnapshotContent as well as VolumeSnapshot from the GCP snapshots, then delete the VolumeSnapshot object. Therefore, the VolumeSnapshotContent remains.

Could you please elaborate more for the purpose?

Reply:
this is for a future PR:
when we need to remove workspace and all associated volume snapshots for it, that could be spread over multiple clusters. 
So I was thinking I could send request to soft delete to all clusters except one that will actually perform hard delete.

Reply:
@jenting there is an example of soft delete usage now in this PR.